chemprop/cli/hyperopt.py

```diff
@@ -1,16 +1,348 @@
+import json
+import logging
+import sys
 from argparse import ArgumentParser, Namespace
+from pathlib import Path
 
+import torch
+from lightning import pytorch as pl
+from lightning.pytorch.loggers import CSVLogger, TensorBoardLogger
+from lightning.pytorch.callbacks import ModelCheckpoint, EarlyStopping
+from ray import tune
+from ray.train import CheckpointConfig, RunConfig, ScalingConfig
+from ray.train.lightning import (RayDDPStrategy, RayLightningEnvironment,
+                                 RayTrainReportCallback, prepare_trainer)
+from ray.train.torch import TorchTrainer
+from ray.tune.schedulers import ASHAScheduler
+
+from chemprop.cli.common import (add_common_args, process_common_args,
+                                 validate_common_args)
+from chemprop.cli.train import (add_train_args, build_datasets, build_model,
+                                build_splits, normalize_inputs,
+                                process_train_args, validate_train_args)
 from chemprop.cli.utils.command import Subcommand
+from chemprop.data import MolGraphDataLoader
+from chemprop.nn import AggregationRegistry
+from chemprop.nn.utils import Activation
+
+
+logger = logging.getLogger(__name__)
+
+AVAILABLE_SPACES = {
```

Comment:
I think this is a really good default. However, It would be good to give the user the option to pass in their own search space as well.

Reply:
Good point! What type of input format do you think we should use?

Reply:
Just leaving a comment as per our discussion today that we are leaning towards supporting a JSON input that contains the dictionary which we translate to the hyperparameter search space.

Reply:
Do we plan to make this available via the CLI?

Reply:
@JacksonBurns We plan to make this available via the CLI, but we haven't figured out an appropriate input format. In previous dev meeting, people think JSON is a good format to use. I tried, but functions are not JSON serializable. I tried many ways to translate the search space dictionary to be JSON serializable, but they all feel hacky. I will implement a version of CLI available through the pickle format. Feel free to chime in on this.

Reply:
So input this as a JSON format would look something like this.

```
{
    "activation":
        {
            "type": "choice",
            "args": ["relu", "gelu", "tanh"],
        },
    "aggregation":
        {
            "type": "choice",
            "args": ["mean", "sum", "max",],
        },
    "aggregation_norm":
        {
            "type": "quniform",
            "kwargs": {"low": 1,
            "high": 200,
            "q": 1,}
        },
    "batch_size":
        {
            "type": "quniform",
            "kwargs": {"low": 5,
            "high": 200,
            "q": 5,}
        },
    "depth":
        {
            "type": "randint",
            "kwargs": {"low": 2,
            "high": 6,}
        },
    "dropout":
        {
            "type": "choice",
            "args": [
                {"type": "choice", "args": [0.0]},
                {"type": "quniform", "kwargs": {"low": 0.05, "high": 0.4, "q": 0.05}},
            ],
        },
    "ffn_hidden_size":
        {
            "type": "quniform",
            "kwargs": {"low": 300, "high": 2400, "q": 100,}
        },
    "ffn_num_layers":
        {
            "type": "randint",
            "kwargs": {"low": 2, "high": 6,}
        },
    "final_lr_ratio":
        {
            "type": "loguniform",
            "kwargs": {"low": 1e-4, "high": 1,}
        },
    "hidden_size":
        {
            "type": "quniform",
            "kwargs": {"low": 300,
            "high": 2400,
            "q": 100,}
        },
    "init_lr_ratio":
        {
            "type": "loguniform",
            "kwargs": {"low": 1e-4,
            "high": 1,}
        },
    "linked_hidden_size":
        {
            "type": "quniform",
            "kwargs": {"low": 300,
            "high": 2400,
            "q": 100,}
        },
    "max_lr":
        {
            "type": "loguniform",
            "kwargs": {"low": 1e-6,
            "high": 1e-2,}
        },
}
```

This JSON corresponds to a python dictionary for search space

```
{
    "activation": tune.choice(["relu", "gelu", "tanh"]),
    "aggregation": tune.choice(["mean", "sum", "max",]),
    "aggregation_norm": tune.quniform(lower=1, upper=200, q=1),
    "batch_size": tune.quniform(lower=5, upper=200, q=5),
    "depth": tune.randint(lower=2, upper=6),
    "dropout": tune.choice([tune.choice([0.0]), tune.quniform(lower=0.05, upper=0.4, q=0.05)]),
    "ffn_hidden_size": tune.quniform(lower=300, upper=2400, q=100),
    "ffn_num_layers": tune.randint(lower=2, upper=6),
    "final_lr_ratio": tune.loguniform(lower=1e-4, upper=1),
    "hidden_size": tune.quniform(lower=300, upper=2400, q=100),
    "init_lr_ratio": tune.loguniform(lower=1e-4, upper=1),
    "linked_hidden_size": tune.quniform(lower=300, upper=2400, q=100),
    "max_lr": tune.loguniform(lower=1e-6, upper=1e-2),
}
```

I prefer the PICKLE format because
- it is more flexible to handle the user input type
- The JSON format requires the same amount of user knowledge on ray tune function as the PICKLE format
- We don't have to maintain our custom JSON format to PICKLE format translator

Reply:
I see two questions here:
1. "How can we accept plaintext input to parameterize the search space for a given hyperparameter?"
2. "More broadly, can we programmatically define a search space using plaintext input?" 

---

For (1), we'd have to design some sort of factory pattern around the `ray.tune` search space API. This actually isn't that hard and requires (i) a mapping from a string key to the corresponding function and (ii) a way to pass parameters to that function. Luckily, [Ray has already provided (i) for us](https://docs.ray.io/en/latest/tune/api/search_space.html):
```python
DISTRIBUTIONS = {
    "uniform": tune.uniform,
    "quniform": tune.quniform,
    "loguniform": tune.loguniform,
    "qloguniform": tune.qloguniform,
    "randn": tune.randn,
    "qrandn": tune.qrandn,
    "randint": tune.randint,
    "qrandint": tune.qrandint,
    "lograndint": tune.lograndint,
    "qlograndint": tune.qlograndint,
    "choice": tune.choice,
    "func": tune.sample_from,    # NOTE: not a good idea
    "grid": tune.grid_search,
}
```
You could also just do `getattr(tune, key)`, but there's less control there in which search spaces we allow. From there, you just pass through the received parameters like so: `func(**params)`. So you would need some sort of plaintext format that enables providing a mapping like this:
```json
{
  "distribution": "uniform",
  "params": {
    "lower": 0.0,
    "upper": 1.0
  }
}
```
> [!Note]
> I'm using JSON here, but you can use YAML, TOML, INI, etc.

Then, inside `hyperopt.py`, you can just create the search space like so:
```python
d = json.load(...)
search_space = DISTRIBUTIONS[d['distribution']](**d['params'])
# or equivalently
search_space = getattr(tune, d['distribution'])(**d['params'])
```

There would need to be some error checking here to validate the input parameters. This can be somewhat cumbersome because we will need to establish, for each hyperparameter, the allowable search space parameterizations. Ex: `"aggregation"` can _only_ be `"choice"` with `params["choices"]` $\subseteq$ `{"max", "norm", "mean", "sum"}` and `"batch_size"` can be any _integer_ distribution with a lower bound $\geq$ 1 _or_ `tune.choice` where `params["choices"]` $\subseteq \mathbb N^n$ 

---

For (2), we would need to establish a mapping from a given hyperparameter key to its actual hyperparameter. We already have this in the form of our `train.py` args and the `update_args_with_config()` method defined below.

Reply:
@davidegraff Thanks for the detailed write-up! Note that using this would not allow a nested search space such as the one below, which is used in our default, unless we make the parsing function parse recursively.

```
"dropout": tune.choice([tune.choice([0.0]), tune.quniform(lower=0.05, upper=0.4, q=0.05)]),
```

Reply:
Yeah that's one of the drawbacks to allowing "fully user-defined search spaces" unless we _also_ define our own text parser. I think that's outside the scope of the chemprop CLI and can thus be limited to (1) our hardcoded search space or (2) a user-implemented hyperoptimization.

Reply:
I decide to keep this part of changes as a separate PR and focus on getting the base case hyperparameter optimization in for now.